[cc1101] Add new cc1101 component

lygris opened the pull request:
# What does this implement/fix?

This PR introduces a new component to support the **Texas Instruments CC1101** Sub-1GHz Transceiver.

**This is a refactored and simplified fork of #6300.** It builds upon the original work but addresses specific hardware quirks and logic issues found during testing. I have numbers, selects, sensors, and etc already coded out and working, but it will make the PR much larger so they are planned for a future PR.

* **Refactored Logic:** Simplified the codebase to use a robust, non-blocking state machine for initialization, ensuring `setup()` does not block the main loop.
* **Stable Reset Sequence:** Implements robust polling of the `CHIP_RDYn` signal (via the MISO pin during `NOP` strobes) to prevent race conditions during the power-on reset sequence.
* **Core Integration:** Designed to work seamlessly with `remote_receiver` and `remote_transmitter` for processing raw OOK/ASK signals (e.g., 433.92MHz remotes).

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- supersedes #6300

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#5614

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [x] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx
- [ ] nRF52840

## Example entry for `config.yaml`:

```yaml
spi:
  clk_pin: GPIO18
  mosi_pin: GPIO23
  miso_pin: GPIO19

cc1101:
  id: cc1101_transceiver
  cs_pin: GPIO5
  frequency: 433.92MHz
  modulation: ASK/OOK
  bandwidth: 203kHz
  symbol_rate: 5000

# Example: Receiving raw data via GDO0
remote_receiver:
  pin: 
       number: GPIO4 # Connected to CC1101 GDO0
       allow_other_uses: true 
  dump: all
  
# Example: Transmitting raw data
remote_transmitter:
  pin:
       number: GPIO5
       allow_other_uses: true 
  carrier_duty_percent: 100%
```

## Checklist:
- [x] The code change is tested and works locally.
- [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
- [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

lygris committed:
Refactor and simpily CC1101 component. Originally forked from PR#6300

lygris committed:
Fixing test configs.

lygris committed:
Merge branch 'dev' into cc1101-core

lygris committed:
remove dependencies.lock

swoboda1337 commented on esphome/components/cc1101/cc1101.h:
```diff
@@ -0,0 +1,132 @@
+  }
+
+ public:
+};
```
This needs to be removed. It won't work with non blocking tx and we are not going to to do this for every remote transmitter action. The user can do it properly like this: https://esphome.io/components/sx127x/#as-a-transmitter

lygris replied:
removed

swoboda1337 commented on tests/test.cc1101.yaml:
This file should be removed

swoboda1337 commented on esphome/components/cc1101/cc1101.cpp:
```diff
@@ -0,0 +1,749 @@
+  if ((value) < (min_value) || (value) > (max_value)) { \
+    ESP_LOGE(TAG, "%s(%d) invalid (%d - %d)", __func__, (int) (value), (int) (min_value), (int) (max_value)); \
+    return; \
+  }
```
Can these be replaced with constexpr functions?

lygris replied:
i could replace them with with with these
```
// Helper for checking enums
template<typename T>
constexpr bool is_enum_valid(T value) {
  // Assumes your enums all have a ::LAST member
  return value < T::LAST;
}

// Helper for checking float ranges
constexpr bool is_float_in_range(float value, float min, float max) {
  return value >= min && value <= max;
}

// Helper for checking int ranges
constexpr bool is_int_in_range(int value, int min, int max) {
  return value >= min && value <= max;
}
```
  
  and the functions calling them would all have to change to have and if statement in them like this
  ```
  void set_frequency(float freq) {
  // You must write the 'if' and 'return' yourself
  if (!is_float_in_range(freq, 0.0, 100.0)) {
    ESP_LOGE(TAG, "Invalid frequency: %f", freq); // Optional: Log it yourself here
    return;
  }
  ```

swoboda1337 replied:
I will take another look at these tomorrow. Would just be good to avoid defines if possible

lygris replied:
I updated them and committed the new version.

swoboda1337 commented on esphome/components/cc1101/cc1101.cpp:
```diff
@@ -0,0 +1,749 @@
+    noInterrupts();  // NOLINT
+#else                // USE_ESP_IDF
+    portDISABLE_INTERRUPTS()
+#endif
```
These interrupt enable/disables should be removed.

lygris replied:
removed

swoboda1337 reviewed (changes requested):
Did a quick first pass and these stood out

lygris committed:
Fix PR review items

lygris committed:
Merge branch 'cc1101-core' of https://github.com/lygris/esphome into cc1101-core

lygris committed:
Merge branch 'dev' into cc1101-core

lygris committed:
Refactor CC1101 component for stability and ESPHome standards

This commit implements several critical improvements to the CC1101 component to align with ESPHome core standards and ensure long-term stability.

Key Changes:
* **Stability:** Replaced `std::deque` with a custom `FixedQueue` ring buffer to eliminate dynamic heap allocation in the main loop, preventing memory fragmentation.
* **HAL:** Migrated from `InternalGPIOPin` to the generic `GPIOPin` abstraction, enabling support for non-ESP platforms (e.g., RP2040, LibreTuya).
* **Architecture:** Updated `const.py` to inherit from `cg.Component` instead of `cg.PollingComponent`, correctly matching the C++ event-driven implementation.
* **Safety:** Added `__attribute__((packed))` to the `CC1101State` struct in `cc1101defs.h` to guarantee strict memory alignment with hardware registers.
* **Cleanup:** Standardized headers in `cc1101pa.h` (added `<cstdint>`, `<cstddef>`) and optimized startup delays to better yield to the system scheduler.

lygris committed:
Merge remote-tracking branch 'upstream/dev' into cc1101-core

lygris committed:
**Code Cleanup and Transmit Fix:**
* **Stability:**
    * Removed the asynchronous `cmd_queue_` mechanism.
    * Fixed missing register addresses and command strobes that were causing TX to be broken.
    * `begin_tx` is now blocking, ensuring the radio has physically entered TX state before `remote_transmitter` begins sending data.
* **Architecture:** Flattened the YAML configuration schema. Nested `tuner:` and `agc:` keys are removed; all options are now top-level configuration items.
* **Code Cleanup:**
    * Renamed C++ setter methods to remove redundant prefixes (e.g., `set_tuner_frequency` -> `set_frequency`).
    * Refactored `__init__.py` to use a `CONFIG_MAP` for cleaner, automated code generation.
    * Updated `const.py` to remove complex nested dictionary structures.
* **HAL Compatibility:** Migrated from `InternalGPIOPin` to generic `GPIOPin` and `gpio_output_pin_schema` to support external GPIO expanders and other platforms.
* **Safety:** Added `__attribute__((packed))` to the `CC1101State` struct to guarantee strict memory alignment with hardware registers.

swoboda1337 commented on esphome/components/cc1101/cc1101.h:
```diff
@@ -0,0 +1,120 @@
+    SET_FREQ_WAIT_RX,
+  };
+
+  void process_cmd_queue_();
```
Can't find this function.

swoboda1337 commented on esphome/components/cc1101/cc1101.cpp:
```diff
@@ -0,0 +1,842 @@
+}
+
+void CC1101Component::begin_tx() {
+  this->send_(Command::TX);
+
+  if (this->gdo0_ != nullptr) {
+#ifdef USE_ESP8266
+    // On ESP8266, we explicitly switch the pin to output for TX
+    this->gdo0_->pin_mode(gpio::FLAG_OUTPUT);
+#endif
+  }
+}
```
I think this has serious issues. If you look at `this->send_` all it does is queue up the command which is handled later in the loop. 

How does transmitting anything even work? It won't switch to transmit until after remote transmitter finishes transmitting!

swoboda1337 replied:
The function `begin_tx` needs to block until its in tx mode. Can this cmd_queue_ stuff get ripped out?

lygris replied:
i'll take a look, I forgot to test transmit because my main use for this is receiving 433MHZ temp sensors.

swoboda1337 replied:
I tested tx on an 8266 with this PR and with my old fork of #6300. My fork works but this PR does nothing at all. Rx works fine on both:
```
external_components:
  - source: github://swoboda1337/components-esphome@main
    components: [ cc1101 ]
    
cc1101:
  id: cc1101_id
  cs_pin: GPIO15
  bandwidth: 200
  frequency: 433920
  
remote_receiver:
  id: rx_id
  pin: GPIO5
  filter: 100us
  idle: 800us
  buffer_size: 2kb

remote_transmitter:
  id: tx_id
  pin: GPIO4
  carrier_duty_percent: 100%
  on_transmit:
    then:
      - cc1101.begin_tx
  on_complete:
    then:
      - cc1101.end_tx

button:
  - platform: template
    name: "Template Button"
    on_press:
      then:
        - remote_transmitter.transmit_raw:
            code: [100, -100, 100, -100, 200, -200, 200, -200, 300, -300, 300, -300, 400, -400, 400, -400, 500, -500, 500, -500, 600, -600, 600, -600, 100, -100, 100, -100, 200, -200, 200, -200, 300, -300, 300, -300, 400, -400, 400, -400, 500, -500, 500, -500, 600, -600, 600, -600, 100, -100, 100, -100, 200, -200, 200, -200, 300, -300, 300, -300, 400, -400, 400, -400, 500, -500, 500, -500, 600, -600, 600, -600]  
```

swoboda1337 reviewed (changes requested):
Does transmit even work?

daveconde commented:
I brought in the latest commit this morning but I’m no longer seeing any pulses coming in from the receiver. I have been having signal strength issues so the issue could be on my side but I figured I’d mention it in case.

lygris committed:
Fixes

lygris commented:
> I brought in the latest commit this morning but I’m no longer seeing any pulses coming in from the receiver. I have been having signal strength issues so the issue could be on my side but I figured I’d mention it in case.

Fixed it, I was in a hurry yesterday and didn't realize. Can you test transmit as well? I've been trying to but i think the boards I got from amazon might be faulty on the transmit side like one of the reviews says.

lygris committed:
Fix component test file

swoboda1337 commented:
I will try transmit again tomorrow

swoboda1337 commented on CODEOWNERS:
```diff
@@ -95,6 +95,7 @@ esphome/components/camera_encoder/* @DT-art1
 esphome/components/canbus/* @danielschramm @mvturnho
 esphome/components/cap1188/* @mreditor97
 esphome/components/captive_portal/* @esphome/core
+esphome/components/cc1101/* @lygris
```
```suggestion
esphome/components/cc1101/* @lygris @gabest11 
```

swoboda1337 commented on tests/components/cc1101/test.esp32-ard.yaml:
Can delete this file, only need IDF

swoboda1337 commented:
I tested it and both rx and tx work now

lygris committed:
Merge branch 'esphome:dev' into cc1101-core

lygris committed:
Fix PR requested changes

lygris committed:
Merge branch 'cc1101-core' of https://github.com/lygris/esphome into cc1101-core

lygris committed:
fix CODEOWNERS

lygris committed:
Fix CODEOWNERS

lygris committed:
Add `set_idle` action for state management

lygris committed:
feat(cc1101): Refactor GDO config with `operation_mode`, align pins, and add `set_idle`

This major refactor revamps how the CC1101 GDO pins are configured, aligning the component with the physical constraints of common 8-pin modules and introducing explicit state control.

Key Changes:

1.  **GDO Configuration Modes:** The configuration is managed by the new `operation_mode` parameter (`DUAL_PIN`, `SINGLE_PIN`) replacing the confusing `gdo0_pin` property. This abstracts complex IOCFG register settings based on the user's wiring setup.
2.  **Pin Alignment:** GDO configuration is explicitly set to align with the chip's I/O capabilities, specifically dedicating the:
    * **GDO0:** Configured for **TX Data Input** in `DUAL_PIN` mode, as it's the only pin capable of acting as an input.
    * **GDO2:** Configured for **RX Data Output** in `DUAL_PIN` mode.
3.  **Cleanup & Removal:** Removes all obsolete code related to:
    * The `gdo0_pin` property (from YAML config and C++ code).
    * The redundant `GDO0_ADC_ID` constant and associated logic.
4.  **New Automation Action:** Adds the `cc1101.set_idle` action, providing necessary control for safely transitioning the chip state, especially during complex TX/RX automations.

Affected Files:
- cc1101/const.py: Added `OPERATION_MODE` enum; removed `CONF_GDO0_PIN` and `CONF_GDO0_ADC_ID`.
- cc1101/__init__.py: Implemented mode configuration logic; registered `cc1101.set_idle` action; removed `CONF_GDO0_PIN` setup.
- cc1101.h: Added public method `set_idle()` and GDO config setters; removed `gdo0_` member and `set_config_gdo0_pin()`.
- cc1101.cpp: Implemented `set_idle()` and GDO config setters; removed all redundant GDO pin handling and member usage.

Sola85 commented:
Thanks a lot for this PR! This works for my rx-only usecase and I'm looking forward to when this is ready!

I did notice one thing though: Prior to the `operation_mode` change, I received data on `GDO0` as expected. However now, no matter which `operation_mode` I set, I only receive on `GDO2` while `GDO0` outputs nothing.

lygris committed:
Merge branch 'dev' into cc1101-core

lygris committed:
refactor(cc1101): Simplify GDO configuration to use universal defaults

Removes the `operation_mode` configuration and explicit GDO pin management from the Python layer. The component now relies on the C++ driver's default initialization, which configures both GDO0 and GDO2 to `0x0D` (Async Serial I/O).

This default configuration universally supports both wiring schemes without requiring specific YAML configuration:
1.  **Single-Pin:** GDO0 acts as a bi-directional data pin.
2.  **Dual-Pin:** GDO0 acts as the TX Input, and GDO2 acts as the RX Output.

Affected files:
- cc1101/const.py: Removed `CONF_OPERATION_MODE` and GDO register constants.
- cc1101/__init__.py: Removed `operation_mode` logic and GDO setters; configuration is now handled entirely by C++ defaults.

swoboda1337 commented on esphome/components/cc1101/__init__.py:
```diff
@@ -0,0 +1,149 @@
+    CC1101Component,
+)
+
+AUTO_LOAD = ["remote_base"]
```
```suggestion
```

swoboda1337 commented on esphome/components/cc1101/__init__.py:
```diff
@@ -0,0 +1,149 @@
+from esphome import automation
+from esphome.automation import maybe_simple_id
+import esphome.codegen as cg
+from esphome.components import remote_base, spi
```
```suggestion
from esphome.components import spi
```

swoboda1337 commented on esphome/components/cc1101/__init__.py:
```diff
@@ -0,0 +1,149 @@
+    .extend(remote_base.REMOTE_TRANSMITTABLE_SCHEMA)
+    .extend(remote_base.RC_SWITCH_RAW_SCHEMA)
+    .extend(remote_base.RC_SWITCH_TRANSMITTER)
+)
```
```suggestion
```

swoboda1337 commented on esphome/components/cc1101/cc1101.h:
```diff
@@ -0,0 +1,120 @@
+#include "esphome/core/hal.h"
+#include "esphome/components/spi/spi.h"
+#include "esphome/core/automation.h"
+#include "esphome/components/remote_base/rc_switch_protocol.h"
```
```suggestion
```

swoboda1337 commented on esphome/components/cc1101/const.py:
```diff
@@ -0,0 +1,151 @@
+
+CODEOWNERS = ["@lygris @gabest11"]
+DEPENDENCIES = ["spi", "button", "remote_receiver"]
+AUTO_LOAD = ["remote_base", "remote_transmitter", "binary_sensor"]
```
```suggestion
DEPENDENCIES = ["spi"]
```

swoboda1337 commented on esphome/components/cc1101/cc1101defs.h:
```diff
@@ -0,0 +1,681 @@
+#include <cinttypes>
+
+namespace esphome {
+namespace cc1101 {
```
```suggestion
namespace esphome::cc1101 {
```

lygris committed:
Decouple from remote_base and simplify dependencies

As requested during review, this commit refactors the `cc1101` component to remove its direct integration with `remote_base`. This simplifies the component into a more generic, low-level driver for the CC1101 transceiver.

Key changes:
- Removed dependencies on `remote_base`, `button`, and `binary_sensor`. The component now only depends on `spi`.
- Updated C++ namespace declarations in `cc1101defs.h` to the modern `namespace esphome::cc1101` style.

This change makes the component more modular, allowing it to serve as a base for various communication protocols without being tied specifically to the `remote_base` implementation.

Sola85 commented:
> > Thanks a lot for this PR! This works for my rx-only usecase and I'm looking forward to when this is ready!
> > I did notice one thing though: Prior to the `operation_mode` change, I received data on `GDO0` as expected. However now, no matter which `operation_mode` I set, I only receive on `GDO2` while `GDO0` outputs nothing.
> 
> Give it another shot, i removed the operation mode config option, the defaults on setup already accomplish the pin settings for both now i believe.

Yup, now the commit description matches my observed behaviour :) only tested rx though.

swoboda1337 committed:
Clean up

swoboda1337 committed:
Clean up

swoboda1337 committed:
Clean up

swoboda1337 committed:
Add more to dump_config

swoboda1337 committed:
Clean up

swoboda1337 committed:
Clean up

swoboda1337 committed:
Fix codeowners

swoboda1337 committed:
Clean up

lygris commented:
looks like it's working fine!

swoboda1337 committed:
Clean up

swoboda1337 committed:
Clean up

swoboda1337 committed:
Clean up

swoboda1337 committed:
Cleanup

swoboda1337 committed:
Rename end_tx

swoboda1337 committed:
Cleanup

swoboda1337 commented:
Tested without open drain it works
```
esphome:
  on_boot:
    - priority: 0
      then:
        - lambda: id(rx_pin)->pin_mode(gpio::FLAG_INPUT);

remote_receiver:
  id: rx_id
  pin: 
    id: rx_pin
    number: GPIO4
    allow_other_uses: true
  filter: 100us
  idle: 800us
  # dump: raw
  buffer_size: 2kb

remote_transmitter:
  id: tx_id
  pin: 
    id: tx_pin
    number: GPIO4
    allow_other_uses: true
  carrier_duty_percent: 100%
  on_transmit:
    then:
      - cc1101.set_idle
      - lambda: 'id(tx_pin)->pin_mode(gpio::FLAG_OUTPUT);'
      - cc1101.begin_tx
  on_complete:
    then:
      - cc1101.set_idle
      - lambda: 'id(tx_pin)->pin_mode(gpio::FLAG_INPUT);'
      - cc1101.begin_rx
```

swoboda1337 committed:
Fix bug

lygris commented:
The chips are crappy versions and as far as i can tell transmit just doesn't work. An amazon review mentioned it and i'm assuming it's true because i could never pick anything up with my rtl-sdr

swoboda1337 committed:
Fix actions and add tests

swoboda1337 committed:
Cleanup

swoboda1337 committed:
Fix c style casts

swoboda1337 committed:
Cleanup

lygris commented:
> @lygris please review the changes but LGTM

LGTM!
